feat/ Pacifica Finance connector implementation

dizpers opened the pull request:
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] Tests all pass
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:

This the implementation of Hummingbot integration with [Pacifica Finance](https://app.pacifica.fi/) (part of [the bounty](https://github.com/hummingbot/hummingbot/issues/7879)). API docs could be found [here](https://docs.pacifica.fi/api-documentation/api).

**Tests performed by the developer**:

* Unit tests.
* Perpetual market making with order placement, cancellation, trades being done and positions being tracked / managed correctly. (**mainnet** and **testnet**)

<details>

<summary>The config used for testing (mainnet)</summary>

```
########################################################
###       Perpetual market making strategy config         ###
########################################################

template_version: 6
strategy: perpetual_market_making

# derivative and token parameters.
derivative: pacifica_perpetual

# Token trading pair for the exchange, e.g. BTC-USDT
market: SOL-USDC

# What leverage to used
leverage: 10

# Position mode to use, hedge mode or one-way position mode.
position_mode: One-way

# How far away from mid price to place the bid order.
# Spread of 1 = 1% away from mid price at that time.
# Example if mid price is 100 and bid_spread is 1.
# Your bid is placed at 99.
bid_spread: 0.05

# How far away from mid price to place the ask order.
# Spread of 1 = 1% away from mid price at that time.
# Example if mid price is 100 and ask_spread is 1.
# Your bid is placed at 101.
ask_spread: 0.05

# Minimum Spread
# How far away from the mid price to cancel active orders
minimum_spread: -100.0

# Time in seconds before cancelling and placing new orders.
# If the value is 60, the bot cancels active orders and placing new ones after a minute.
order_refresh_time: 1.0

# The spread (from mid price) to defer order refresh process to the next cycle.
# (Enter 1 to indicate 1%), value below 0, e.g. -1, is to disable this feature - not recommended.
order_refresh_tolerance_pct: 0.0

# Size of your bid and ask order.
order_amount: 1.0

# long position take profit spread
long_profit_taking_spread: 0.0

# short position take profit spread
short_profit_taking_spread: 0.0

# Spread from position entry price to place a stop-loss order to close position
stop_loss_spread: 1.0

# Time to wait before refreshing a stop loss order that has not been executed
time_between_stop_loss_orders: 60.0

# Spread to include in stop loss orders covering possible price slippages in the market
stop_loss_slippage_buffer: 0.5

# Price band ceiling.
price_ceiling: -1.0

# Price band floor.
price_floor: -1.0

# Number of levels of orders to place on each side of the order book.
order_levels: 1

# Increase or decrease size of consecutive orders after the first order (if order_levels > 1).
order_level_amount: 0

# Order price space between orders (if order_levels > 1).
order_level_spread: 1.0

# How long to wait before placing the next order in case your order gets filled.
filled_order_delay: 60

# Whether to enable order optimization mode (true/false).
order_optimization_enabled: false

# The depth in base asset amount to be used for finding top ask (for order optimization mode).
ask_order_optimization_depth: 0

# The depth in base asset amount to be used for finding top bid (for order optimization mode).
bid_order_optimization_depth: 0

# The price source (current_market/external_market/custom_api).
price_source: current_market

# The price type (mid_price/last_price/last_own_trade_price/best_bid/best_ask).
price_type: mid_price

# An external exchange name (for external exchange pricing source).
price_source_derivative:

# A trading pair for the external exchange, e.g. BTC-USDT (for external exchange pricing source).
price_source_market:

# An external api that returns price (for custom_api pricing source).
price_source_custom_api:

# An interval time in second to update the price from custom api (for custom_api pricing source).
custom_api_update_interval: 5.0

# Use user provided orders to directly override the orders placed by order_amount and order_level_parameter
# This is an advanced feature and user is expected to directly edit this field in config file
# Below is an sample input, the format is a dictionary, the key is user-defined order name, the value is a list which includes buy/sell, order spread, and order amount
# order_override:
#   order_1: [buy, 0.5, 100]
#   order_2: [buy, 0.75, 200]
#   order_3: [sell, 0.1, 500]
# Please make sure there is a space between : and [
order_override:
```
</details>

<details>

<summary>The config used for testing (testnet)</summary>

```
########################################################
###       Perpetual market making strategy config         ###
########################################################

template_version: 6
strategy: perpetual_market_making

# derivative and token parameters.
derivative: pacifica_perpetual_testnet

# Token trading pair for the exchange, e.g. BTC-USDT
market: SOL-USDC

# What leverage to used
leverage: 10

# Position mode to use, hedge mode or one-way position mode.
position_mode: One-way

# How far away from mid price to place the bid order.
# Spread of 1 = 1% away from mid price at that time.
# Example if mid price is 100 and bid_spread is 1.
# Your bid is placed at 99.
bid_spread: 1.0

# How far away from mid price to place the ask order.
# Spread of 1 = 1% away from mid price at that time.
# Example if mid price is 100 and ask_spread is 1.
# Your bid is placed at 101.
ask_spread: 1.0

# Minimum Spread
# How far away from the mid price to cancel active orders
minimum_spread: -100.0

# Time in seconds before cancelling and placing new orders.
# If the value is 60, the bot cancels active orders and placing new ones after a minute.
order_refresh_time: 3.0

# The spread (from mid price) to defer order refresh process to the next cycle.
# (Enter 1 to indicate 1%), value below 0, e.g. -1, is to disable this feature - not recommended.
order_refresh_tolerance_pct: 0.0

# Size of your bid and ask order.
order_amount: 0.75

# long position take profit spread
long_profit_taking_spread: 1.0

# short position take profit spread
short_profit_taking_spread: 1.0

# Spread from position entry price to place a stop-loss order to close position
stop_loss_spread: 1.0

# Time to wait before refreshing a stop loss order that has not been executed
time_between_stop_loss_orders: 60.0

# Spread to include in stop loss orders covering possible price slippages in the market
stop_loss_slippage_buffer: 0.5

# Price band ceiling.
price_ceiling: -1.0

# Price band floor.
price_floor: -1.0

# Number of levels of orders to place on each side of the order book.
order_levels: 1

# Increase or decrease size of consecutive orders after the first order (if order_levels > 1).
order_level_amount: 0

# Order price space between orders (if order_levels > 1).
order_level_spread: 1.0

# How long to wait before placing the next order in case your order gets filled.
filled_order_delay: 60

# Whether to enable order optimization mode (true/false).
order_optimization_enabled: false

# The depth in base asset amount to be used for finding top ask (for order optimization mode).
ask_order_optimization_depth: 0

# The depth in base asset amount to be used for finding top bid (for order optimization mode).
bid_order_optimization_depth: 0

# The price source (current_market/external_market/custom_api).
price_source: current_market

# The price type (mid_price/last_price/last_own_trade_price/best_bid/best_ask).
price_type: mid_price

# An external exchange name (for external exchange pricing source).
price_source_derivative:

# A trading pair for the external exchange, e.g. BTC-USDT (for external exchange pricing source).
price_source_market:

# An external api that returns price (for custom_api pricing source).
price_source_custom_api:

# An interval time in second to update the price from custom api (for custom_api pricing source).
custom_api_update_interval: 5.0

# Use user provided orders to directly override the orders placed by order_amount and order_level_parameter
# This is an advanced feature and user is expected to directly edit this field in config file
# Below is an sample input, the format is a dictionary, the key is user-defined order name, the value is a list which includes buy/sell, order spread, and order amount
# order_override:
#   order_1: [buy, 0.5, 100]
#   order_2: [buy, 0.75, 200]
#   order_3: [sell, 0.1, 500]
# Please make sure there is a space between : and [
order_override:

```
</details>

**Tips for QA testing**:

<details>

<summary>scripts/pacifica_candles_test.py</summary>

```python3
from typing import Dict

import pandas as pd

from hummingbot.connector.connector_base import ConnectorBase
from hummingbot.data_feed.candles_feed.candles_factory import CandlesFactory
from hummingbot.data_feed.candles_feed.data_types import CandlesConfig
from hummingbot.strategy.script_strategy_base import ScriptStrategyBase


class PacificaCandlesExample(ScriptStrategyBase):
    pacifica_1m_candles = CandlesFactory.get_candle(
        CandlesConfig(connector="pacifica_perpetual", trading_pair="BTC-USDC", interval="1m", max_records=200)
    )
    pacifica_1h_candles = CandlesFactory.get_candle(
        CandlesConfig(connector="pacifica_perpetual", trading_pair="BTC-USDC", interval="1h", max_records=200)
    )

    markets = {}

    def __init__(self, connectors: Dict[str, ConnectorBase]):
        super().__init__(connectors)
        self.pacifica_1m_candles.start()
        self.pacifica_1h_candles.start()

    @property
    def all_candles_ready(self):
        return all([self.pacifica_1m_candles.ready, self.pacifica_1h_candles.ready])

    async def on_stop(self):
        self.pacifica_1m_candles.stop()
        self.pacifica_1h_candles.stop()

    def format_status(self) -> str:
        if not self.all_candles_ready:
            return "Candles not ready yet."
        lines = []
        for candles in [self.pacifica_1h_candles, self.pacifica_1m_candles]:
            candles_df = candles.candles_df.copy()
            candles_df["timestamp"] = pd.to_datetime(candles_df["timestamp"], unit="s")
            lines.extend([f"Candles: {candles.name} | Interval: {candles.interval}"])
            lines.extend(["    " + line for line in candles_df.tail().to_string(index=False).split("\n")])
            lines.extend([""])
        return "\n".join(lines)
```
</details>

<details>

<summary>scripts/test_pacifica_candles_standalone.py</summary>

```python3
#!/usr/bin/env python3
"""
Simple test runner for Pacifica candles feed.
Tests both 1m and 1h intervals to verify REST and WebSocket functionality.
"""
import asyncio
import logging

from hummingbot.data_feed.candles_feed.pacifica_perpetual_candles import PacificaPerpetualCandles

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)


async def test_pacifica_candles():
    """Test Pacifica candles feed with 1m interval"""
    logger.info("=" * 80)
    logger.info("Testing Pacifica Perpetual Candles Feed")
    logger.info("=" * 80)

    # Initialize candles feed for BTC-USDC 1m
    logger.info("\n1. Initializing 1m candles feed for BTC-USDC...")
    candles_1m = PacificaPerpetualCandles(
        trading_pair="BTC-USDC",
        interval="1m",
        max_records=10  # Keep it small for quick testing
    )

    try:
        # Start the network (fetches historical data and subscribes to WebSocket)
        logger.info("2. Starting network connection...")
        await candles_1m.start_network()

        # Wait for candles to be ready (fill with historical data)
        logger.info("3. Waiting for candles to be ready...")
        wait_time = 0
        max_wait = 30  # 30 seconds max wait

        while not candles_1m.ready and wait_time < max_wait:
            await asyncio.sleep(1)
            wait_time += 1
            if wait_time % 5 == 0:
                logger.info(f"   Waiting... ({wait_time}s, {len(candles_1m._candles)}/{candles_1m.max_records} candles loaded)")

        if candles_1m.ready:
            logger.info(f"✅ Candles feed is ready! ({len(candles_1m._candles)} candles loaded)")

            # Display candle data
            df = candles_1m.candles_df
            logger.info("\n4. Latest candles data:")
            logger.info(f"\n{df.tail(5).to_string()}")

            # Verify data integrity
            logger.info("\n5. Data validation:")
            logger.info(f"   - Total candles: {len(df)}")
            logger.info(f"   - Latest close price: ${df.iloc[-1]['close']:.2f}")
            logger.info(f"   - Latest volume: {df.iloc[-1]['volume']:.6f}")
            logger.info(f"   - Price range (last 5): ${df.tail(5)['low'].min():.2f} - ${df.tail(5)['high'].max():.2f}")

            # Check if prices are reasonable (BTC should be in a reasonable range)
            latest_close = df.iloc[-1]['close']
            if 10000 < latest_close < 200000:
                logger.info(f"   ✅ Price validation passed (${latest_close:.2f} is within reasonable BTC range)")
            else:
                logger.warning(f"   ⚠️  Unusual price detected: ${latest_close:.2f}")

            # Test WebSocket by waiting a bit and checking for updates
            logger.info("\n6. Testing WebSocket real-time updates...")
            logger.info("   Waiting 65 seconds for new candle...")
            initial_count = len(candles_1m._candles)

            for i in range(13):  # 13 * 5 = 65 seconds (just over 1 minute)
                await asyncio.sleep(5)
                current_count = len(candles_1m._candles)
                if current_count > initial_count:
                    logger.info(f"   ✅ New candle received! ({current_count} candles now)")
                    latest = candles_1m.candles_df.iloc[-1]
                    logger.info(f"      Latest: ${latest['close']:.2f} | Vol: {latest['volume']:.6f}")
                    break
                else:
                    logger.info(f"   ... waiting ({(i + 1) * 5}s)")

            logger.info("\n" + "=" * 80)
            logger.info("✅ TEST COMPLETED SUCCESSFULLY")
            logger.info("=" * 80)

        else:
            logger.error(f"❌ Candles feed failed to initialize within {max_wait}s")
            logger.error(f"   Only {len(candles_1m._candles)}/{candles_1m.max_records} candles loaded")

    except Exception as e:
        logger.error(f"❌ Error during test: {e}", exc_info=True)

    finally:
        # Clean up
        logger.info("\n7. Stopping candles feed...")
        await candles_1m.stop_network()
        logger.info("   Stopped.")


if __name__ == "__main__":
    asyncio.run(test_pacifica_candles())

```
</details>

* **New Dependencies**: This PR introduces new package dependencies. You must run `./install` to update your environment.
* **Key Configuration**: You only need to provide:
    * **Agent Wallet Public Key**: Derived automatically or can be provided if needed. (better to provide)
    * **Agent Wallet Private Key**: The private key of your dedicated agent wallet.
    * **User Wallet Public Key**: The public key (address) of your main wallet connected to the Pacifica Finance application.
* **Candles feed** logic has been provided and should be tested (e.g. using the script above)
* Pacifica Perpetual was registered as a **rate oracle**. But keep in mind that the exchange only supports USDC as a quote asset.
* There's need to perform test like you did for other connectors in https://github.com/hummingbot/hummingbot/pull/7930

More about agent wallets: https://docs.pacifica.fi/api-documentation/api/signing/api-agent-keys
You could create an Agent Wallet (API Key) here: https://app.pacifica.fi/apikey

More information could be discovered here: https://github.com/hummingbot/hummingbot-site/pull/486

dizpers committed:
(feat) Implementation of Pacifica Finance connector

dizpers commented:
Update: Rebased against upstream/development and force-pushed.
New Feature: Added rate oracle support for Pacifica (covered with unit-tests)

Note: Pacifica only supports USDC as quote token

Also: looks like my forced push kicked @nikspz merge commit (he was adding latest development changes)

nikspz committed:
Merge branch 'development' into feat/implement-pacifica-connector

nikspz committed:
Merge branch 'development' into feat/implement-pacifica-connector

nikspz reviewed (commented):
@dizpers Could you please fix the warning message I reported to you earlier

commit 12775432b5fe1babeac009277dfc91173240efab

```
2026-02-06 18:29:15,902 - 6502 - hummingbot.core.trading_core - INFO - Markets recorder initialized with database: conf_perpetual_market_making_1
2026-02-06 18:29:15,903 - 6502 - hummingbot.core.trading_core - INFO - Creating the clock with tick size: 1.0
2026-02-06 18:29:15,904 - 6502 - hummingbot.core.trading_core - INFO - Clock started successfully
2026-02-06 18:29:15,905 - 6502 - hummingbot.core.trading_core - INFO - 'perpetual_market_making' strategy execution started.
2026-02-06 18:29:15,905 - 6502 - hummingbot.core.trading_core - INFO - Strategy perpetual_market_making started successfully
2026-02-06 18:29:16,270 - 6502 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_derivative.PacificaPerpetualDerivative - INFO - Network status has changed to NetworkStatus.CONNECTED. Starting networking...
2026-02-06 18:29:16,270 - 6502 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_derivative.PacificaPerpetualDerivative - INFO - Starting network...
2026-02-06 18:29:16,270 - 6502 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_derivative.PacificaPerpetualDerivative - INFO - api_config_key: 
2026-02-06 18:29:16,611 - 6502 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_derivative.PacificaPerpetualDerivative - INFO - Using existing API Config Key: 5tyQwEK7_CXo789MjGwN18sP76ujpJ
2026-02-06 18:29:16,775 - 6502 - hummingbot.core.rate_oracle.rate_oracle - INFO - Network status has changed to NetworkStatus.CONNECTED. Starting networking...
2026-02-06 18:29:16,808 - 6502 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_derivative.PacificaPerpetualDerivative - INFO - Leverage for SOL-USDC successfully set to 3.
2026-02-06 18:29:16,305 - 6502 - hummingbot.core.data_type.order_book_tracker - INFO - Initialized order book for SOL-USDC. 1/1 completed.
2026-02-06 18:29:16,502 - 6502 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_derivative.PacificaPerpetualDerivative - ERROR - TRADE FEES UPDATED
2026-02-06 18:29:16,537 - 6502 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_derivative.PacificaPerpetualDerivative - ERROR - Failed to fetch last fee payment (no data): {'success': True, 'data': [], 'has_more': False}
2026-02-06 18:29:16,870 - 6502 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_derivative.PacificaPerpetualDerivative - ERROR - Failed to fetch last fee payment (no data): {'success': True, 'data': [], 'has_more': False}
2026-02-06 18:29:17,056 - 6502 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_user_stream_data_source.PacificaPerpetualUserStreamDataSource - INFO - Subscribed to private account and orders channels
2026-02-06 18:29:17,157 - 6502 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_api_order_book_data_source.PacificaPerpetualAPIOrderBookDataSource - INFO - Subscribed to public order book and trade channels...
2026-02-06 18:29:24,004 - 6502 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - Changing position mode to ONEWAY succeeded.
2026-02-06 18:29:24,334 - 6502 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_derivative.PacificaPerpetualDerivative - INFO - Leverage for SOL-USDC successfully set to 3.
2026-02-06 18:29:25,005 - 6502 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (SOL-USDC) Creating 1 LIMIT bid orders at (Size, Price): ['0.15 SOL, 80.79 USDC'] to OPEN position.
2026-02-06 18:29:25,008 - 6502 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (SOL-USDC) Creating 1  LIMIT ask orders at (Size, Price): ['0.15 SOL, 81.11 USDC'] to OPEN position.
2026-02-06 18:29:25,538 - 6502 - hummingbot.connector.client_order_tracker - INFO - Created LIMIT SELL order HBOTSSLUC64a2619f5b3a0a59d2fb3d0 for 0.15 to OPEN a SOL-USDC position at 81.11.
2026-02-06 18:29:25,551 - 6502 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1770377365.0, "type": "OrderType.LIMIT", "trading_pair": "SOL-USDC", "amount": "0.15", "price": "81.11", "order_id": "HBOTSSLUC64a2619f5b3a0a59d2fb3d0", "creation_timestamp": 1770377365.0, "exchange_order_id": "4316892468", "leverage": 3, "position": "OPEN", "event_name": "SellOrderCreatedEvent", "event_source": "pacifica_perpetual"}
2026-02-06 18:29:25,551 - 6502 - hummingbot.connector.client_order_tracker - INFO - Created LIMIT BUY order HBOTBSLUC64a2619f5a8caa59d2fb3d0 for 0.15 to OPEN a SOL-USDC position at 80.79.
2026-02-06 18:29:25,561 - 6502 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1770377365.0, "type": "OrderType.LIMIT", "trading_pair": "SOL-USDC", "amount": "0.15", "price": "80.79", "order_id": "HBOTBSLUC64a2619f5a8caa59d2fb3d0", "creation_timestamp": 1770377365.0, "exchange_order_id": "4316892473", "leverage": 3, "position": "OPEN", "event_name": "BuyOrderCreatedEvent", "event_source": "pacifica_perpetual"}
2026-02-06 18:30:00,652 - 6502 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_derivative.PacificaPerpetualDerivative - ERROR - Failed to fetch last fee payment (no data): {'success': True, 'data': [], 'has_more': False}
2026-02-06 18:30:14,071 - 6502 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1770377414.0, "order_id": "HBOTBSLUC64a2619f5a8caa59d2fb3d0", "exchange_order_id": "4316892473", "event_name": "OrderCancelledEvent", "event_source": "pacifica_perpetual"}
2026-02-06 18:30:14,071 - 6502 - hummingbot.connector.client_order_tracker - INFO - Successfully canceled order HBOTBSLUC64a2619f5a8caa59d2fb3d0.
2026-02-06 18:30:14,080 - 6502 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1770377414.0, "order_id": "HBOTSSLUC64a2619f5b3a0a59d2fb3d0", "exchange_order_id": "4316892468", "event_name": "OrderCancelledEvent", "event_source": "pacifica_perpetual"}
2026-02-06 18:30:14,080 - 6502 - hummingbot.connector.client_order_tracker - INFO - Successfully canceled order HBOTSSLUC64a2619f5b3a0a59d2fb3d0.
2026-02-06 18:30:15,103 - 6502 - hummingbot.client.hummingbot_application - ERROR - MQTT is already stopped!
```

<img width="2479" height="1499" alt="Image" src="https://github.com/user-attachments/assets/fa34b21f-6f58-4e44-a48d-1a3b0cfafeec" />

nikspz reviewed (commented):
commit 12775432b5fe1babeac009277dfc91173240efab

2nd issue when you stop v1 perpetualMM strategy bot flooded with 
PacificaPerpetualUserStreamDataSource - Error sending ping to Pacifica WebSocket (See log file for stack trace dump) 


[logs_conf_perpetual_market_making_1.log](https://github.com/user-attachments/files/25126751/logs_conf_perpetual_market_making_1.log)

```
2026-02-06 18:37:45,852 - 6867 - hummingbot.core.trading_core - INFO - Strategy stopped successfully
2026-02-06 18:37:45,855 - 6867 - hummingbot.core.trading_core - INFO - Clock stopped successfully
2026-02-06 18:37:55,698 - 6867 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_user_stream_data_source.PacificaPerpetualUserStreamDataSource - WARNING - Error sending ping to Pacifica WebSocket
Traceback (most recent call last):
  File "/home/etozhe/pacifica7907/hummingbot/connector/derivative/pacifica_perpetual/pacifica_perpetual_user_stream_data_source.py", line 103, in _ping_loop
    await ws.send(WSJSONRequest(payload={"op": "ping"}))
  File "/home/etozhe/pacifica7907/hummingbot/core/web_assistant/ws_assistant.py", line 63, in send
    await self._connection.send(request)
  File "/home/etozhe/pacifica7907/hummingbot/core/web_assistant/connections/ws_connection.py", line 56, in send
    self._ensure_connected()
    ~~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/etozhe/pacifica7907/hummingbot/core/web_assistant/connections/ws_connection.py", line 79, in _ensure_connected
    raise RuntimeError("WS is not connected.")
RuntimeError: WS is not connected.
2026-02-06 18:37:55,911 - 6867 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_api_order_book_data_source.PacificaPerpetualAPIOrderBookDataSource - WARNING - Error sending ping to Pacifica WebSocket
Traceback (most recent call last):
  File "/home/etozhe/pacifica7907/hummingbot/connector/derivative/pacifica_perpetual/pacifica_perpetual_api_order_book_data_source.py", line 239, in _ping_loop
    await ws.send(ping_request)
  File "/home/etozhe/pacifica7907/hummingbot/core/web_assistant/ws_assistant.py", line 63, in send
    await self._connection.send(request)
  File "/home/etozhe/pacifica7907/hummingbot/core/web_assistant/connections/ws_connection.py", line 56, in send
    self._ensure_connected()
    ~~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/etozhe/pacifica7907/hummingbot/core/web_assistant/connections/ws_connection.py", line 79, in _ensure_connected
    raise RuntimeError("WS is not connected.")
RuntimeError: WS is not connected.
2026-02-06 18:38:29,972 - 6867 - hummingbot.connector.derivative.pacifica_perpetual.pacifica_perpetual_user_stream_data_source.PacificaPerpetualUserStreamDataSource - WARNING - Error sending ping to Pacifica WebSocket
Traceback (most recent call last):
  File "/home/etozhe/pacifica7907/hummingbot/connector/derivative/pacifica_perpetual/pacifica_perpetual_user_stream_data_source.py", line 103, in _ping_loop
    await ws.send(WSJSONRequest(payload={"op": "ping"}))
  File "/home/etozhe/pacifica7907/hummingbot/core/web_assistant/ws_assistant.py", line 63, in send
    await self._connection.send(request)
  File "/home/etozhe/pacifica7907/hummingbot/core/web_assistant/connections/ws_connection.py", line 56, in send
    self._ensure_connected()
    ~~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/etozhe/pacifica7907/hummingbot/core/web_assistant/connections/ws_connection.py", line 79, in _ensure_connected
    raise RuntimeError("WS is not connected.")
RuntimeError: WS is not connected.
```

<img width="2479" height="1498" alt="image" src="https://github.com/user-attachments/assets/827ea196-a61b-4d4a-a0b1-98feaf6159b9" />

nikspz commented:
commit 12775432b5fe1babeac009277dfc91173240efab


Notes: `Hedge` mode not supported now ⬅

- Test performed:
    - Connected API keys successfully: ok ✅
      -  connected with API Agent Keys (Public and Secret) + Wallet address ✅
    - Run command balance and to see the correct tokens that have assets: ok ✅
    - Created /started perpetualMM strategy successfully ✅
    - Funding payment: issue reported above (https://github.com/hummingbot/hummingbot/pull/7907#pullrequestreview-3762585474) ⬅
    - Check status successfully ✅
    - Position retrieval: ok ✅
    - Order cancellation: ok ✅
      - flooded with warning on stop command https://github.com/hummingbot/hummingbot/pull/7907#pullrequestreview-3762638609
    - Check data integrity for status, order_book --live: ok ✅
    - Order filled events: ok ✅
    - Fees matched with exchange ✅
    - History: ok ✅
    - Data aggregate: ok ✅

Pending: longrun test

nikspz reviewed (commented):
commit 1277543

@dizpers Could you please check this issue on v2 strategies

steps:
0. could use in conf/controllers
[conf_pmm_miste-pacifica2.yml](https://github.com/user-attachments/files/25245429/conf_pmm_miste-pacifica2.yml)

1. create --script-config v2_with_controllers
2. start --script v2_with_controllers.py --conf conf_pmm_miste-pacifica2.yml

Actual:
This issue showing on start of v2 strategy and in the logs during strategy


[logs_conf_pmm_miste-pacifica2.log](https://github.com/user-attachments/files/25245421/logs_conf_pmm_miste-pacifica2.log)

```
2026-02-12 05:11:41,978 - 888 - hummingbot.data_feed.market_data_provider - ERROR - Error fetching prices from pacifica_perpetual: expected a sequence of length 64 (got 0)
Traceback (most recent call last):
  File "/home/etozhe/pacifica7907/hummingbot/data_feed/market_data_provider.py", line 179, in update_rates_task
    connector_instance = self._non_trading_connectors[connector]
                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^
  File "/home/etozhe/pacifica7907/hummingbot/core/data_type/common.py", line 118, in __missing__
    self[key] = self.default_value_factory(key)
                ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^
  File "/home/etozhe/pacifica7907/hummingbot/data_feed/market_data_provider.py", line 310, in _create_non_trading_connector
    connector = connector_class(**init_params)
  File "/home/etozhe/pacifica7907/hummingbot/connector/derivative/pacifica_perpetual/pacifica_perpetual_derivative.py", line 81, in __init__
    super().__init__(balance_asset_limit=balance_asset_limit, rate_limits_share_pct=rate_limits_share_pct)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/etozhe/pacifica7907/hummingbot/connector/perpetual_derivative_py_base.py", line 31, in __init__
    super().__init__(balance_asset_limit, rate_limits_share_pct)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/etozhe/pacifica7907/hummingbot/connector/exchange_py_base.py", line 70, in __init__
    self._auth: AuthBase = self.authenticator
                           ^^^^^^^^^^^^^^^^^^
  File "/home/etozhe/pacifica7907/hummingbot/connector/derivative/pacifica_perpetual/pacifica_perpetual_derivative.py", line 89, in authenticator
    return PacificaPerpetualAuth(
        agent_wallet_public_key=self.agent_wallet_public_key,
        agent_wallet_private_key=self.agent_wallet_private_key,
        user_wallet_public_key=self.user_wallet_public_key,
    )
  File "/home/etozhe/pacifica7907/hummingbot/connector/derivative/pacifica_perpetual/pacifica_perpetual_auth.py", line 21, in __init__
    self.keypair = Keypair.from_bytes(base58.b58decode(self.agent_wallet_private_key))
                   ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: expected a sequence of length 64 (got 0)
```

<img width="2480" height="1507" alt="image" src="https://github.com/user-attachments/assets/67f65c63-6fda-471f-b97b-a641bd828956" />

nikspz committed:
Merge branch 'development' into feat/implement-pacifica-connector

dizpers committed:
(feature) Dynamic rate limits; close ping loop on a new connection

dizpers committed:
(feature) Lazy keypari init

nikspz committed:
Merge branch 'development' into feat/implement-pacifica-connector

nikspz committed:
Merge branch 'development' into feat/implement-pacifica-connector

nikspz committed:
Merge branch 'development' into feat/implement-pacifica-connector

nikspz reviewed (commented):
Note: pacifica_perpetual connector not available in US
Note: Only Oneway mode available for pacifica_perpetual connector

commit 4365ddf66b1a0ed28dd9bde9aba7a3a5cb1974ff

- Connected API keys successfully: ok ✅
  -  connected with API Agent Keys (Public and Secret) + Wallet address ✅
- Run command balance and to see the correct tokens that have assets: ok ✅
- Created /started v1 perpetualMM strategy successfully ✅
- available to start v2 bot successfully now ✅
- got same issue as here https://github.com/hummingbot/hummingbot/pull/7920#pullrequestreview-3828514291 for pmm_mister bot for pacifica_perpetual
  - bot accumulated big short position ❌
  - <img width="2480" height="1535" alt="image" src="https://github.com/user-attachments/assets/3b065654-fdf3-4885-9fc6-14b35b1318aa" />
  - also had issue with network on middle of the run ⏸
    - bot was able to resolve, but still accumulated  big short position as described above
- from that going to check on another v2 strategies since connector only support Oneway

logs:
[19feb.zip](https://github.com/user-attachments/files/25428385/19feb.zip)

Could you please check market order implementation also @dizpers 
[logs_macd-bb.log](https://github.com/user-attachments/files/25428782/logs_macd-bb.log)





pending: retest on other (directional) v2 strategy

nikspz committed:
Merge branch 'development' into feat/implement-pacifica-connector

dizpers committed:
(fix) add missing argument required for market orders

dizpers commented:
@nikspz 

## Fix: Market orders missing required `slippage_percent` field

### Problem

Market orders on `pacifica_perpetual` connector fail with:
```
POST /api/v1/orders/create_market → 400: Json deserialize error: missing field `slippage_percent`
```
The Pacifica API [requires](https://docs.pacifica.fi/api-documentation/api/rest-api/orders/create-market-order) `slippage_percent` for market orders, but the connector's `_place_order()` method doesn't include it.

### Fix

- Added `MARKET_ORDER_MAX_SLIPPAGE = "5"` constant (5% — same as Hyperliquid and Derive connectors)
- Added `data["slippage_percent"]` to market order payload in `_place_order()`
- Added `test_place_market_order` unit test


---

## Separate Issue: Position Accumulation (same as in #7920)

> **This is a strategy-level issue, not a connector bug. Should be posted as a separate issue.**

During QA testing of the `pmm_mister` strategy on `pacifica_perpetual`, the bot accumulated a large short position (~8.91 SOL). Log analysis shows:

| | OPEN fills | CLOSE fills |
|----|------|-------|
| BUY | 7 | 15 |
| SELL | **42** | 5 |

**Root cause**: In ONEWAY mode, `pmm_mister`'s position executors operate independently -- one opens a LONG, another simultaneously opens a SHORT. On the exchange, these net into a single position. When a SELL OPEN flips the net position from LONG to SHORT, the BUY executor's take-profit (`reduce_only=true`) is rejected by the exchange:
```
422: Invalid reduce-only order side: must be opposite to position side
```
The take-profit retries all fail (up to 10x), and the strategy keeps opening new SELL positions → shorts accumulate.

**The connector correctly translates `PositionAction.CLOSE → reduce_only: true`.** The fix needs to go into the position executor / strategy layer to coordinate executors in ONEWAY mode (where multiple independent position executors can't assume isolated positions because the exchange nets them).

dizpers committed:
Merge branch 'development' into feat/implement-pacifica-connector

nikspz committed:
Merge branch 'development' into feat/implement-pacifica-connector

nikspz committed:
Merge branch 'development' into feat/implement-pacifica-connector

nikspz committed:
Update conf_fee_overrides_TEMPLATE.yml

nikspz reviewed (approved):
Note: US Region is restricted for API trading
Note: Hedge mode not supported now

- commit 1e14dd1
    - Test performed:
        - Cloned and installed PR7907 ✅
        - connected pacifica_perpetual successfully ✅
        - checked balance successfully ✅
        - created/started v1 perpetualMM successfully ✅
            - review orders created/ cancelled successfully ✅
            - status: ok ✅
            - position retrieval: ok
            - order_book —live: ok ✅
            - limit filled orders: ok ✅
        - created/started v2 strategy bb_v1 successfully ✅
            - reviewed issue with market order fixed ✅
            - market orders filled successully ✅
            - longrun: ok ✅

rapcmia committed:
Merge branch 'development' into feat/implement-pacifica-connector

nikspz committed:
Merge branch 'development' into feat/implement-pacifica-connector
